### PR TITLE
XRT-1354 branch

### PR DESCRIPTION
**Pull request by SteveOss**
Scheduler optimisation

### CODE REVIEW

**Comment by SteveOss** [comment]
Mix of IOT data utility functions update and scheduler re-write. Scheduler update is extensive so needs reviewing as a whole not a diff.

**Comment by SteveOss** [comment]
@BinduRao2018 Looks like some sonarqube issue with the pr-merge. Will investigate.

**Comment by SteveOss** [comment]
@BinduRao2018 New gcovr release was breaking SonarQube scan. Have pinned version to fix.

**Review by BinduRao2018 — COMMENTED** [review]
In Scheduler.c,
  Thread - iot_scheduler_thread():  shouldn't the cond_wait only on 'Running'? If the state is 'STOPPED' will this wait for 'scheduler->schd_time', which shouldn't be the case?

test_data_const_pointer() utest - added comment inline

File: src/c/utests/data/data.c [hunk 1]
```diff
@@ -3854,6 +3877,18 @@ static void test_data_const_string (void)
...
+static void test_data_const_pointer (void)
+{
+  static iot_data_static_t block;
+  static const void * ptr = &block;
+  iot_data_t * data = iot_data_alloc_const_pointer (&block, ptr);
+  CU_ASSERT (iot_data_pointer (data) == ptr)
+  CU_ASSERT (data == IOT_DATA_STATIC (block))
+  CU_ASSERT (iot_data_is_static (data))
+  iot_data_free (data);
+  iot_data_free (data);
```
[the remark this answers is not in the document]
Comment (SteveOss) [reply src/c/utests/data/data.c:3889]: It's static const data, so freeing has no effect, this was to check that.

**Comment by BinduRao2018** [comment]
recheck

**Pull request merged**